WordPress/gutenberg
# Enforce valid function names in the packages/block-library/src/*/*.php files

## What?
This PR is focused on implementing a new PHPCS sniff, namely `Gutenberg.NamingConventions.ValidBlockLibraryFunctionName`.
Fixes https://github.com/WordPress/gutenberg/issues/52769.

## Why?
All the PHP function names within the `./packages/block-library/src/*/*.php` files should start with one of the following prefixes:

- `block_core_<folder_name>`
- `render_block_core_<folder_name>`
- `register_block_core_<folder_name>`

Here, `<folder_name>` represents the name of the folder where the associated file is situated. The folder name is converted to lowercase, and any characters other than letters and digits are substituted with underscores.

The introduction of this new PHPCS sniff is needed to ensure strict adherence to this naming pattern.

## How?
The proposed `ValidBlockLibraryFunctionNameSniff` is configured to examine PHP function names contained within the `./packages/block-library/src/*/*.php` files.

Whenever a function name deviates from the naming convention listed above, the PHPCS sniff will add an error to the PHPCS report.

## Testing Instructions
1. Make sure that CI jobs pass.
2. Open the `packages/block-library/src/calendar/index.php` file in your preferred code editor.
3. Add a new PHP function to the file:
```php
/**
 * A test function.
 */
function block_core_calendar_() {

}
```
4. Run the linter (`npm run lint:php` or `composer run lint`, based on your environment setup).
5. Notice the linter error displayed in the console:
```
The function name "block_core_calendar_()" is invalid because PHP function names in this file should start with one of the following prefixes: "block_core_calendar", "render_block_core_calendar", "register_block_core_calendar".
```
6. Rename the php function to `block_core_clndr`.
7. Rerun the linter.
8. Notice the linter error displayed in the console:
```
The function name "block_core_clndr()" is invalid because PHP function names in this file should start with one of the following prefixes: "block_core_calendar", "render_block_core_calendar", "register_block_core_calendar".
```
9. Rename the php function to `block_core_calendar`.
10. Rerun the linter.
11. Confirm the absence of linter errors in the console.
12. [Optional] Perform testing with other allowed prefixes, like `render_block_core_calendar` and `register_block_core_calendar`.
